feat: generate dbt tests using datapilot

saravmajestic opened the pull request:
## Overview

### Requirement
Allow user to use datapilot to generate dbt tests for model/columns using NLP

### Solution
- Added `Custom test` button along with dbt generic tests in add test flow
- Show datapilot panel when the above button is clicked with option to let user enter requirement
- Datapilot will generate sql based custom tests or macros from dbt_utils/expectations if available in dbt project
- If it is yaml suggestion, allow user to insert to schema directly
- if it is sql based, user will manually create sql files 

### Screenshot/Demo
![image](https://github.com/AltimateAI/vscode-dbt-power-user/assets/1147025/b10ee82d-ccf3-40c2-9463-f29175524f35)

### How to test
- Click Custom test button in add test button flow for model/column
- In the datapilot panel, enter your prompt. Ex: `make order_id not to be null`
- Datapilot will suggest `not_null` dbt test for the column
- Click Insert button
- This should add not_null test to the order_id column and show save button
- regression - query analysis
    - check skip followup for query change 
- regression - doc generator

## Checklist

- [x] I have run this code and it appears to resolve the stated issue
- [ ] `README.md` updated and added information about my change

saravmajestic commented on src/services/dbtTestService.ts:
```diff
@@ -168,4 +175,121 @@ export class DbtTestService {
+    });
+  }
+
+  public async getTestsForCurrentModel() {
```
moved from src/services/docGenService.ts

saravmajestic commented on src/webview_provider/datapilotPanel.ts:
```diff
@@ -107,74 +108,79 @@ export class DataPilotPanel extends AltimateWebviewProvider {
-            error: (err as Error).message,
-          });
-        }
+      case "queryAnalysis:explain":
```
used `handleSyncRequestFromWebview` instead of using try catch implementation for each command

saravmajestic commented on src/webview_provider/newDocsGenPanel.ts:
```diff
@@ -130,49 +126,37 @@ export class NewDocsGenPanel
-            args.test as TestMetaData,
-            args.model as string,
-          ),
+        this.handleSyncRequestFromWebview(
```
used `handleSyncRequestFromWebview` instead of repeated try catch blocks

mdesmet commented on src/dbt_client/dbtCoreIntegration.ts:
```diff
@@ -909,6 +909,33 @@ export class DBTCoreProjectIntegration
 
+  findPackageVersion(packageName: string) {
+    if (!this.packagesInstallPath || !packageName) {
+      return;
```
Is this not an error?

saravmajestic replied:
fixed

mdesmet commented on src/dbt_client/dbtCoreIntegration.ts:
```diff
@@ -909,6 +909,33 @@ export class DBTCoreProjectIntegration
+      "dbt_project.yml",
+    );
+    if (!existsSync(dbtProjectYmlFilePath)) {
+      return;
```
Is this not an error? this means package is not installed.

saravmajestic replied:
This is not an error. If the file does not exist, then user does not have this package installed and we dont have to use it.

mdesmet commented on src/dbt_client/dbtCoreIntegration.ts:
```diff
@@ -909,6 +909,33 @@ export class DBTCoreProjectIntegration
+      encoding: "utf-8",
+    });
+    if (!fileContents) {
+      return;
```
Error?

saravmajestic replied:
fixed

mdesmet commented on src/dbt_client/dbtCoreIntegration.ts:
```diff
@@ -909,6 +909,33 @@ export class DBTCoreProjectIntegration
+      uniqueKeys: false,
+      maxAliasCount: -1,
+    });
+    return parsedConfig?.version;
```
if version is not defined is also an error.

saravmajestic replied:
fixed

mdesmet commented on src/dbt_client/dbtIntegration.ts:
```diff
@@ -346,6 +346,7 @@ export interface DBTProjectIntegration extends Disposable {
   getCatalog(): Promise<Catalog>;
   getDebounceForRebuildManifest(): number;
   getBulkSchema(nodes: DBTNode[]): Promise<Record<string, DBColumn[]>>;
+  findPackageVersion(packageName: string): string | undefined;
```
```suggestion
  findPackageVersion(packageName: string): string;
```
What will you do when it doesn't return string?

saravmajestic replied:
Current requirement does not expect the version to be always available. If version is not returned, we can assume that the package is not installed and wont use it. I have updated it to return string or throw error, so the caller will handle this as per the need

mdesmet commented on webview_panels/src/modules/dataPilot/DataPilotProvider.tsx:
```diff
@@ -41,46 +32,21 @@ const DataPilotProvider = ({
-    } as DatapilotQueryAnalysisChat;
+  const handleIncomingDatapilotMessage = (request: Partial<DataPilotChat>) => {
+    panelLogger.info("datapilot incoming message", request);
+    const id = request.id ?? crypto.randomUUID();
```
How can request.id be empty? is that not a bug?

saravmajestic replied:
In case of doc generation, id will be passed from doc editor panel to webview provider and then to datapilot panel. In other cases, id is optional and will be created/used only in datapilot. So this is not a required field

mdesmet replied:
I think it would be better to have it always there in that case to keep things simple, but not blocking for this PR.

mdesmet reviewed (commented):
I have slight concerns on how much we go to check the active text editor. I think this will be a headache to fix if we add unsaved changes handlers. Can't we refactor this to only retrieve the project once and pass as reference to the messages?

saravmajestic commented on src/altimate.ts:
```diff
@@ -113,6 +113,13 @@ export interface QueryAnalysisRequest {
 
+export interface CreateDbtTestRequest {
+  session_id: string;
+  model: DocsGenerateModelRequestV2["dbt_model"];
```
fixed

saravmajestic commented:
> I have slight concerns on how much we go to check the active text editor. I think this will be a headache to fix if we add unsaved changes handlers. Can't we refactor this to only retrieve the project once and pass as reference to the messages?

@mdesmet i have fixed the active editor changes for the dbt test creation flow and updated other places where current active file is enough.

mdesmet commented on src/dbt_client/dbtCloudIntegration.ts:
```diff
@@ -832,6 +832,10 @@ export class DBTCloudProjectIntegration
   }
 
+  findPackageVersion(packageName: string) {
+    return "";
```
I didn't think of dbt cloud, if on cloud we can never get the version and that is not an exception then the return value should indeed be string | undefined. Core integration should still throw an error if it cannot detect the version as this indicates a problem.

mdesmet commented on src/dbt_client/dbtCoreIntegration.ts:
```diff
@@ -909,6 +909,40 @@ export class DBTCoreProjectIntegration
+      maxAliasCount: -1,
+    });
+    if (!parsedConfig?.version) {
+      throw new Error(`Missing version in ${packageName}/dbt_project.yml`);
```
maybe use full path, otherwise we need to guess which project in a multi project setup.

mdesmet commented on src/webview_provider/newDocsGenPanel.ts:
```diff
@@ -130,147 +126,118 @@ export class NewDocsGenPanel
+        this.handleSyncRequestFromWebview(
+          syncRequestId,
+          async () => {
+            const columnsFromSources = await this.queryManifestService
+              .getProject()
```
get project from message?

mdesmet commented on src/webview_provider/newDocsGenPanel.ts:
```diff
@@ -130,147 +126,118 @@ export class NewDocsGenPanel
+        this.handleSyncRequestFromWebview(
+          syncRequestId,
+          async () => {
+            const columns = await this.queryManifestService
```
get project from message?

mdesmet commented on src/webview_provider/newDocsGenPanel.ts:
```diff
@@ -130,147 +126,118 @@ export class NewDocsGenPanel
+        this.handleSyncRequestFromWebview(
+          syncRequestId,
+          async () => {
+            const sources = this.queryManifestService.getSourcesInProject(
```
get project from message?

saravmajestic replied:
all these commands are triggered based on the documentation of current model that is open in editor. do we still need to handle that from message?

mdesmet reviewed (commented):
Please fix signature of dbtIntegration `findPackageVersion`, other stuff is not blocking.
